File: src/main/kotlin/com/nazarethlabs/codex/service/note/DuplicateNoteService.kt
package com.nazarethlabs.codex.service.note

import com.intellij.openapi.diagnostic.thisLogger
import com.intellij.openapi.project.Project
import com.nazarethlabs.codex.dto.Note
import com.nazarethlabs.codex.helper.FileHelper
import com.nazarethlabs.codex.helper.NoteNameHelper
import com.nazarethlabs.codex.repository.NoteStorageRepository

class DuplicateNoteService {
    private val createNoteService = CreateNoteService()

    fun duplicate(
        project: Project,
        note: Note,
    ) {
        try {
            val content = FileHelper.readText(note.filePath)
            val extension = getFileExtension(note.filePath)
            val newTitle = generateDuplicateTitle(note.title)
            createNoteService.createWithContent(project, newTitle, extension, content)
        } catch (e: Exception) {
            thisLogger().error("Failed to duplicate note: ${note.title}", e)
        }
    }

    private fun getFileExtension(filePath: String): String {
        val ext = filePath.substringAfterLast('.', "")
        return if (ext.isNotEmpty()) ".$ext" else ""
    }

    private fun generateDuplicateTitle(baseTitle: String): String {
        val notes = NoteStorageRepository.getInstance().getAllNotes()
        return NoteNameHelper.generateDuplicateName(baseTitle, notes)
    }
}

File: src/service/DuplicateNoteService.ts
import { Note } from '../dto/Note';
import { NoteRepository } from '../repository/NoteRepository';
import { CreateNoteService } from './CreateNoteService';
import { FileHelper } from '../helper/FileHelper';

export class DuplicateNoteService {
  private readonly repository: NoteRepository;
  private readonly createNoteService: CreateNoteService;

  constructor(repository: NoteRepository, createNoteService: CreateNoteService) {
    this.repository = repository;
    this.createNoteService = createNoteService;
  }

  async duplicate(note: Note): Promise<Note> {
    const content = FileHelper.readText(note.filePath);
    const newTitle = this.generateDuplicateTitle(note.title);
    return this.createNoteService.createWithTitle(newTitle, content);
  }

  private generateDuplicateTitle(originalTitle: string): string {
    const existing = this.repository.getAllNotes().map((n) => n.title);
    let index = 1;
    let title = `${originalTitle} (${index})`;
    while (existing.includes(title)) {
      index++;
      title = `${originalTitle} (${index})`;
    }
    return title;
  }
}
